File: IT/Multithreading/Esercizi/Python/esercizio20.py
# file esercizio20.py

from __future__ import print_function
from multiprocessing import Process, current_process, Value, Semaphore
from random import randint
from time import sleep, time

class Barrier:
	def __init__(self, n):
		self.__n = n;
		self.__count = Value('i', 0)
		self.__mutex = Semaphore(1)
		self.__barrier = Semaphore(0)

	def wait(self):
		self.__mutex.acquire()
		self.__count.value += 1
		self.__mutex.release()

		if (self.__count.value == self.__n):
			self.__barrier.release()

		self.__barrier.acquire()
		self.__barrier.release()

def target(barrier):
	sleep(randint(3, 8))
	barrier.wait()

	# stampa l'istante in millisecondi

	print("{} {}".format(int(round(time()*1000)), current_process().name))

if __name__ == "__main__":
	nProcs = 3
	barrier = Barrier(nProcs)

	procs = [Process(target=target, args=(barrier,)) for i in range(nProcs)]

	for p in procs:
		p.start()

	for p in procs:
		p.join()

File: IT/Multithreading/Esercizi/Java/Esercizio20.java
/* file Esercizio20.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.Semaphore;

class Barrier {
	private int n;
	private int count;
	private Semaphore mutex;
	private Semaphore barrier;

	public Barrier(int n) {
		this.n = n;
		this.count = 0;
		this.mutex = new Semaphore(1);
		this.barrier = new Semaphore(0);
	}

	public void await() {
		try {
			mutex.acquire();
			count++;
			mutex.release();

			if (this.count == this.n) {
				barrier.release();
			}

			barrier.acquire();
			barrier.release();
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
}

class Target implements Runnable {
	private Barrier barrier;

	public Target(Barrier barrier) {
		this.barrier = barrier;
	}

	@Override
	public void run() {
		try {
			int pause = ThreadLocalRandom.current().nextInt(3, 8+1);
			Thread t = Thread.currentThread();
			t.sleep(pause * 1000);
			// System.out.println(System.currentTimeMillis() + " " + t.getName() + " arriva alla barriera.");
			barrier.await();

			// stampa l'istante in millisecondi

			System.out.println(System.currentTimeMillis() + " " + t.getName());
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
}

public class Esercizio20 {
	public static void main(String [] args) {
		final int N_THREADS = 3;
		Barrier barrier = new Barrier(N_THREADS);

		Thread [] threads = new Thread[N_THREADS];

		for (int i = 0; i < N_THREADS; i++) {
			threads[i] = new Thread(new Target(barrier));
		}

		for (Thread t : threads) {
			t.start();
		}
	}
}

